### ISSUE TITLE
x2sys_cross: Refactor to get rid of temporary files and have consistent table-like output behavior

### ISSUE THREAD

**Issue by seisman** [GenericMappingTools/pygmt#3160]
_Originally posted by @weiji14 in https://github.com/GenericMappingTools/pygmt/issues/2730#issuecomment-2001898646_

I had a look at refactoring `x2sys_cross` to use virtualfiles instead of temporary files, but it's a little tricky because:

- Input: Cannot pass in virtualfiles as input as mentioned at https://github.com/GenericMappingTools/pygmt/pull/546#discussion_r459366392 and https://github.com/GenericMappingTools/gmt/issues/3717, since GMT doesn't support virtualfiles to X2SYS modules
- Output: The `virtualfile_to_dataset` method from #3083 was able to produce a `pandas.DataFrame` output, **but** the column names were missing. The logic for handling `x2sys_cross`'s output is actually complicated:

    https://github.com/GenericMappingTools/pygmt/blob/bcbbcad1ffca6f73fcc8ebc2872f228aaaa2266e/pygmt/src/x2sys_cross.py#L231-L250

Important things to handle are:
1. Datetime columns need to be parsed correctly as `datetime64` dtype
2. `x2sys_cross` may output multi-segment parts (see https://docs.generic-mapping-tools.org/6.5/supplements/x2sys/x2sys_cross.html#remarks) when multiple tracks are passed in and `-Qe` (external COEs) is selected. Unsure how this is handled in GMT virtualfiles (note that we actually just merge all the multi-segments into one table when pandas.DataFrame output is selected, output to file will preserve the segments though).
3. Last two column names can either be `z_X`/`z_M` or `z_1`/`z_2` depending on whether trackvalues/`-Z` argument is set.

It should be possible to handle 1 and 3 somehow, but I'm not so sure about 2 since it will involve checking how GMT outputs virtualfiles in `x2sys_cross`. We'll need to do some careful checking to ensure the refactoring doesn't modify the output and makes it incorrect.

**Comment by seisman** [comment]
> * Input: Cannot pass in virtualfiles as input as mentioned at [Wrap x2sys_init and x2sys_cross #546 (comment)](https://github.com/GenericMappingTools/pygmt/pull/546#discussion_r459366392) and [Passing in virtual files into the supplementary x2sys modules gmt#3717](https://github.com/GenericMappingTools/gmt/issues/3717), since GMT doesn't support virtualfiles to X2SYS modules

Nothing we can do on the PyGMT side. On the GMT side, I guess no one will touch the `x2sys_cross` source code in the near future.

> * Output: The `virtualfile_to_dataset` method from [clib: Add virtualfile_to_dataset method for converting virtualfile to a dataset #3083](https://github.com/GenericMappingTools/pygmt/pull/3083) was able to produce a `pandas.DataFrame` output, **but** the column names were missing. 

In #3182, we now parse the colume names from the 2nd row of the output. So the column names should work fine.

https://github.com/GenericMappingTools/pygmt/blob/ef71a9d68b3b9cd98d0655a029c2b56c89329a07/pygmt/src/x2sys_cross.py#L223-L225

> Important things to handle are:
> 
> 1. Datetime columns need to be parsed correctly as `datetime64` dtype

We now call `pd.to_datetime` to convert the columns to `datetime`, so the issue should be solved:
https://github.com/GenericMappingTools/pygmt/blob/ef71a9d68b3b9cd98d0655a029c2b56c89329a07/pygmt/src/x2sys_cross.py#L252

> 3. Last two column names can either be `z_X`/`z_M` or `z_1`/`z_2` depending on whether trackvalues/`-Z` argument is set.

Since we're parsing column names from the file header, the column names now should automatically set. This can be verified by checking our existing `x2sys_cross` tests, in which `z_X`/`z_M` or `z_1`/`z_2` are used in different cases.

> 2. `x2sys_cross` may output multi-segment parts (see https://docs.generic-mapping-tools.org/6.5/supplements/x2sys/x2sys_cross.html#remarks) when multiple tracks are passed in and `-Qe` (external COEs) is selected. Unsure how this is handled in GMT virtualfiles (note that we actually just merge all the multi-segments into one table when pandas.DataFrame output is selected, output to file will preserve the segments though).

> It should be possible to handle 1 and 3 somehow, but I'm not so sure about 2 since it will involve checking how GMT outputs virtualfiles in x2sys_cross. We'll need to do some careful checking to ensure the refactoring doesn't modify the output and makes it incorrect.

Currently, the whole PyGMT project is lacking support of multi-segment files. The `GMT_DATASET` container has all the information about multi-segment files, but all the information are lost when we convert it to pandas.DataFrame. So we need to find another data structure in the Python  for holding multi-segment files. And this will be a long-term goal.

In summary, I think we already fix most of the issues mentioned above, so it can be closed. Feel free to reopen if you disagree.
